### PR TITLE
Add activity log timeline

### PR DESCRIPTION
**Pull request by Nitemaeric**
## Context

We need to show the timeline of activities support users have performed on sets of claims, such as processing payments, samplings, and clawbacks.

We encapsulate these events in `Claims::ClaimActivity` records.

## Changes proposed in this pull request

- Rename `Claims::ActivityLogsController` to `Claims::ClaimActivitiesController` to align with the `Claims::ClaimActivity` class.
- Add some example activities to be displayed in the Claims timeline.

## Guidance to review

- The Sampling record is to show the example of how an activity associated with multiple files would look like. This code is intended to be updated in further PRs.

### Claim Activities

CleanShot 2024-12-23 at 17 43 04@2x

### Payments - Claims sent to ESFA example

CleanShot 2024-12-23 at 17 37 10@2x

### Sampling - Claims sent to Providers example

CleanShot 2024-12-23 at 17 38 58@2x

### CODE REVIEW

File: app/controllers/claims/support/claims/claim_activities_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    @pagy, @claim_activities = pagy(Claims::ClaimActivity.order(created_at: :desc))
+
+    authorize [:claims, @claim_activities]
+  end
+
+  def show
+    @claim_activity = Claims::ClaimActivity.find(params[:id])
+
+    if @claim_activity.sampling_uploaded?
+      @pagy, @provider_samplings = pagy(@claim_activity.record.provider_samplings.joins(:provider).order(providers: { name: :asc }))
```
Comment (JamieCleare2525) [inline app/controllers/claims/support/claims/claim_activities_controller.rb:12]: Might be worth moving some of this to a scope

Comment (Nitemaeric) [reply app/controllers/claims/support/claims/claim_activities_controller.rb]: Resolved here: [fixup! Show Claim Activities as a timeline](https://github.com/DFE-Digital/itt-mentor-services/pull/1261/commits/0bd1bb81540fc756514e39c6c305b00a8649c4e3)

File: app/models/claims/claim_activity.rb [hunk 2]
```diff
@@ -20,11 +20,18 @@
...
+    payment_request_delivered: "payment_request_delivered",
+    payment_response_uploaded: "payment_response_uploaded",
+    sampling_uploaded: "sampling_uploaded",
+    sampling_response_uploaded: "sampling_response_uploaded",
+    clawback_request_delivered: "clawback_request_delivered",
+    clawback_response_uploaded: "clawback_response_uploaded",
   }
+
+  delegate :full_name, to: :user, prefix: true, allow_nil: true
+>>>>>>> eada890e (Show Claim Activities as a timeline)
```
Comment (JamieCleare2525) [inline app/models/claims/claim_activity.rb:36]: rebase issue?

Comment (Nitemaeric) [reply app/models/claims/claim_activity.rb]: Fixed.

File: app/views/claims/support/claims/claim_activities/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  <h1 class="govuk-heading-l"><%= Claims::ClaimActivity.human_attribute_name("action.#{@claim_activity.action}") %></h1>
+
+  <p class="govuk-body"><%= t(".byline", user: @claim_activity.user_full_name, time: l(@claim_activity.created_at, format: :date_at_time)) %></p>
+
+  <h2 class="govuk-heading-m"><%= t(".files") %></h2>
+
+  <%= govuk_list type: :bullet do %>
+    <% case @claim_activity.action %>
+    <% when "payment_request_delivered", "clawback_request_delivered" %>
+      <li><%= govuk_link_to "Claims sent to ESFA", @claim_activity.record.csv_file %></li>
```
Comment (JamieCleare2525) [inline app/views/claims/support/claims/claim_activities/show.html.erb:19]: Do these links need to be `no_visited_state: true`?

File: app/views/claims/support/claims/claim_activities/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    <% when "sampling_uploaded" %>
+      <% @provider_samplings.each do |provider_sampling| %>
+        <li><%= govuk_link_to provider_sampling.provider.name, provider_sampling.csv_file %></li>
```
Comment (JamieCleare2525) [inline app/views/claims/support/claims/claim_activities/show.html.erb:22]: Do these links need to be `no_visited_state: true`?

File: app/views/claims/support/claims/claim_activities/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                          <%= image_tag "icon-document.svg", class: "app-timeline__document-link__icon" %> <%= provider_sampling.provider.name %>
+                        <% end %>
+                      </li>
+                    <% end %>
+                  </ul>
+
+                  <% if provider_samplings.offset(5).any? %>
+                    <p class="govuk-body"><%= t(".showing", count: provider_samplings.limit(5).length, total_count: provider_samplings.count) %></p>
+
+                    <%= govuk_link_to t(".view_all_files"), claims_support_claims_claim_activity_path(claim_activity) %>
```
Comment (JamieCleare2525) [inline app/views/claims/support/claims/claim_activities/index.html.erb:53]: Do these links need to be `no_visited_state: true`?

File: app/views/claims/support/claims/claim_activities/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                      <% end %>
+                    </li>
+                  </ul>
+                <% when "sampling_uploaded" %>
+                  <% provider_samplings = claim_activity.record.provider_samplings %>
+
+                  <ul class="app-timeline__documents">
+                    <% provider_samplings.limit(5).each do |provider_sampling| %>
+                      <li class="app-timeline__document-item">
+                        <%= govuk_link_to provider_sampling.csv_file, class: "app-timeline__document-link" do %>
```
Comment (JamieCleare2525) [inline app/views/claims/support/claims/claim_activities/index.html.erb:43]: Do these links need to be `no_visited_state: true`?

File: app/views/claims/support/claims/claim_activities/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                  <%= t(".byline", user: claim_activity.user_full_name, time: l(claim_activity.created_at, format: :date_at_time)) %>
+                </p>
+              </div>
+
+              <div class="app-timeline__description">
+                <% case claim_activity.action %>
+                <% when "payment_request_delivered", "clawback_request_delivered" %>
+                  <ul class="app-timeline__documents">
+                    <li class="app-timeline__document-item">
+                      <%= govuk_link_to claim_activity.record.csv_file, class: "app-timeline__document-link" do %>
```
Comment (JamieCleare2525) [inline app/views/claims/support/claims/claim_activities/index.html.erb:32]: Do these links need to be `no_visited_state: true`?

File: app/views/claims/support/claims/claim_activities/show.html.erb (hunk 3 above)
Comment (Nitemaeric) [reply app/views/claims/support/claims/claim_activities/show.html.erb:19]: I don't think they should, no. It is valuable to know if a file has been downloaded or not.

File: app/views/claims/support/claims/claim_activities/show.html.erb (hunk 4 above)
Comment (Nitemaeric) [reply app/views/claims/support/claims/claim_activities/show.html.erb:22]: I don't think they should, no. It is valuable to know if a file has been downloaded or not.

File: app/views/claims/support/claims/claim_activities/index.html.erb (hunk 5 above)
Comment (Nitemaeric) [reply app/views/claims/support/claims/claim_activities/index.html.erb:53]: I think this is also valuable to know if the details page has been visited or not. It's not like an index page.

File: app/views/claims/support/claims/claim_activities/index.html.erb (hunk 6 above)
Comment (Nitemaeric) [reply app/views/claims/support/claims/claim_activities/index.html.erb:43]: Same here. I don't think they should. It is valuable to know if a file has been downloaded or not.

File: app/views/claims/support/claims/claim_activities/index.html.erb (hunk 7 above)
Comment (Nitemaeric) [reply app/views/claims/support/claims/claim_activities/index.html.erb:32]: Same again here.

**Comment by Nitemaeric** [comment]
https://github.com/DFE-Digital/itt-mentor-services/blob/68584582fd3913a4e2ec2aab5e852e465b24d7b6/spec/factories/claims/samplings.rb#L11 sets `sampling` to `nil`. This prevents an infinite loop when creating a `provider_sampling` or `claims_sampling`. The `sampling` on `provider_sampling` is set once the `provider_sampling` or `sampling` is saved.

**Review by JamieCleare2525 — APPROVED** [review]

**Pull request merged**